Jenkinsfile

```diff
@@ -0,0 +1,183 @@
+def entrySet(aMap) {
+  aMap.collect {
+    k, v -> [key: k, value: v]
+  }
+}
+
+/** Write capabilities to JSON file
+ *
+ * @param desiredCapabilities capabilities to include in the file
+*/
+def writeCapabilities(desiredCapabilities) {
+    def defaultCapabilities = [
+        build: env.BUILD_TAG,
+        public: 'public restricted'
+    ]
+    def capabilities = defaultCapabilities.clone()
+    capabilities.putAll(desiredCapabilities)
+    def json = JsonOutput.toJson([capabilities: capabilities])
+    writeFile file: 'capabilities.json', text: json
+}
+
+/** Run Tox
+ *
+ * @param environment test environment to run
+*/
+def runTox(environment) {
+  try {
+    wrap([$class: 'AnsiColorBuildWrapper']) {
+      withCredentials([[
+        $class: 'StringBinding',
+        credentialsId: 'SAUCELABS_API_KEY',
+        variable: 'SAUCELABS_API_KEY']]) {
+        withEnv(["PYTEST_ADDOPTS=${PYTEST_ADDOPTS} " +
+          "--base-url=${PYTEST_BASE_URL} " +
+          "--driver=SauceLabs " +
+          "--variables=capabilities.json " +
+          "--color=yes"]) {
+          sh "tox -e ${environment}"
+        }
+      }
+    }
+  } catch(err) {
+    currentBuild.result = 'FAILURE'
+    throw err
+  } finally {
+    dir('results') {
+      stash environment
+    }
+  }
+}
+
+/** Send a notice to #fxtest-alerts on irc.mozilla.org with the build result
+ *
+ * @param result outcome of build
+*/
+def ircNotification(result) {
+  def nick = "fxtest${BUILD_NUMBER}"
+  def channel = '#fx-test-alerts'
+  result = result.toUpperCase()
+  def message = "Project ${JOB_NAME} build #${BUILD_NUMBER}: ${result}: ${BUILD_URL}"
+  node {
+    sh """
+        (
+        echo NICK ${nick}
+        echo USER ${nick} 8 * : ${nick}
+        sleep 5
+        echo "JOIN ${channel}"
+        echo "NOTICE ${channel} :${message}"
+        echo QUIT
+        ) | openssl s_client -connect irc.mozilla.org:6697
+    """
+  }
+}
+
+stage('Checkout') {
+  node {
+    timestamps {
+      deleteDir()
+      checkout scm
+      stash 'workspace'
+    }
+  }
+}
+
+stage('Lint') {
+  node {
+    timestamps {
+      deleteDir()
+      unstash 'workspace'
+      sh 'tox -e flake8'
+    }
+  }
+}
+
+def builders = [:]
+for (entry in entrySet(environments)) {
+  def environment = entry.key
+  def capabilities = entry.value
+  builders[(environment)] = {
+    node {
+      timeout(time: 1, unit: 'HOURS') {
+        timestamps {
+          deleteDir()
+          unstash 'workspace'
+          writeCapabilities(capabilities)
+          withCredentials([[
+            $class: 'FileBinding',
+            credentialsId: 'AMO_VARIABLES',
+            variable: 'VARIABLES']]) {
+            withEnv(["PYTEST_ADDOPTS=--variables=${VARIABLES}"]) {
+              runTox(environment)
+            }
+          }
+        }
+      }
+    }
+  }
+}
+
+try {
+  stage('Test') {
+    parallel builders
+  }
+} catch(err) {
+  currentBuild.result = 'FAILURE'
+  ircNotification(currentBuild.result)
+  mail(
```

Comment:
the default vals approach may not be a bad idea.  then you could just do something like: notificationMail()  and/or notificationIRC().  obviously the recipients would differ from team-to-team, but each project would have its own Jenkinsfile anyway.  the benefit here, would be to simplify the Jenkinsfile by replacing a whole bunch of bash stuff with a simple one-liner.